dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,154 @@
 namespace dwave::optimization {
 
+std::vector<ssize_t> make_concatenate_shape(std::vector<ArrayNode*> array_ptrs, ssize_t axis) {
+    // One or more arrays must be given
+    if (array_ptrs.size() < 1) {
+        throw std::invalid_argument("need at least one array to concatenate");
+    }
+
+    for (auto it = std::next(array_ptrs.begin()), stop = array_ptrs.end(); it != stop; it++) {
+
+        // Arrays must have the same number of dimensions
+        if ((*std::prev(it))->ndim() != (*it)->ndim()) {
+            throw std::invalid_argument(
+                    "all the input arrays must have the same number of dimensions," +
+                    std::string(" but the array at index ") +
+                    std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                    " has " + std::to_string((*std::prev(it))->ndim()) +
+                    " dimension(s) and the array at index " +
+                    std::to_string(std::distance(array_ptrs.begin(), it)) +
+                    " has " +
+                    std::to_string((*it)->ndim()) +
+                    " dimension(s)");
+        }
+
+        // Array shapes must be the same except for on the concatenation axis
+        for (ssize_t i = 0; i < (*it)->ndim(); i++) {
+            if (i != axis) {
+                if ( (*std::prev(it))->shape()[i] != (*it)->shape()[i] ) {
+                    throw std::invalid_argument(
+                            "all the input array dimensions for the concatenation" +
+                            std::string(" axis must match exactly, but along dimension ") +
+                            std::to_string(i) + ", the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                            " has size " +
+                            std::to_string((*std::prev(it))->shape()[i]) +
+                            " and the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), it)) +
+                            " has size " +
+                            std::to_string((*it)->shape()[i]));
+                }
+            }
+        }
+    }
+
+    // Axis must be in range 0..ndim-1
+    // We can do this check on the first input array since
+    // we at this point know they all have the same shape
+    if (!(0 <= axis && axis < array_ptrs.at(0)->ndim())) {
+        throw std::invalid_argument(
+                "axis " +
+                std::to_string(axis) +
+                std::string(" is out of bounds for array of dimension ") +
+                std::to_string(array_ptrs.at(0)->ndim()));
+    }
+
+
+    // The shape of the input arrays, which will be the
+    // same except for possibly on the concatenation axis
+    std::span<const ssize_t> shape0 = array_ptrs.at(0)->shape();
+    std::vector<ssize_t> shape;
+    shape.reserve(shape0.size());
+    std::copy(shape0.begin(), shape0.end(), std::back_inserter(shape));
```

Comment:
Can just do
```c++
    std::vector<ssize_t> shape(shape0.begin(), shape0.end());
```
because span returns a random access iterator the constructor should be smart enough to to the reserve etc.

Reply:
fixed